test/e2e/common/expansion.go

```diff
@@ -306,6 +307,90 @@ var _ = framework.KubeDescribe("Variable Expansion", func() {
 		// Pod should fail
 		testPodFailSubpath(f, pod)
 	})
+
+	/*
+		    Testname: var-expansion-subpath-lifecycle
+		    Description: Verify that a subpath expansion can be modified during the lifecycle of a container.
+			1.	invalid subpathexpr blocks a container from running
+			2.	changes to the expansion take effect on container restart
+			3.	successful expansion of the subpathexpr isn't required for volume cleanup
+
+	*/
+	It("should succeed in altering lifecycle of container [Feature:VolumeSubpathEnvExpansion][NodeAlphaFeature:VolumeSubpathEnvExpansion][Slow]", func() {
+
+		podName := "var-expansion-" + string(uuid.NewUUID())
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        podName,
+				Labels:      map[string]string{"name": podName},
+				Annotations: map[string]string{"notmysubpath": "mypath"},
+			},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:  "dapi-container",
+						Image: imageutils.GetE2EImage(imageutils.BusyBox),
+						Env: []v1.EnvVar{
+							{
+								Name:  "POD_NAME",
+								Value: "foo",
+							},
+							{
+								Name: "ANNOTATION",
+								ValueFrom: &v1.EnvVarSource{
+									FieldRef: &v1.ObjectFieldSelector{
+										APIVersion: "v1",
+										FieldPath:  "metadata.annotations['mysubpath']",
+									},
+								},
+							},
+						},
+						VolumeMounts: []v1.VolumeMount{
+							{
+								Name:        "workdir1",
+								MountPath:   "/logscontainer",
+								SubPathExpr: "$(ANNOTATION)/$(POD_NAME)",
+							},
+						},
+					},
+				},
+				RestartPolicy: v1.RestartPolicyNever,
+				Volumes: []v1.Volume{
+					{
+						Name: "workdir1",
+						VolumeSource: v1.VolumeSource{
+							EmptyDir: &v1.EmptyDirVolumeSource{},
+						},
+					},
+				},
+			},
+		}
+
+		By("creating the pod with failed condition")
+		pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(pod)
+		Expect(err).ToNot(HaveOccurred(), "while creating pod")
+
+		err = framework.WaitTimeoutForPodRunningInNamespace(f.ClientSet, pod.Name, pod.Namespace, framework.PodStartShortTimeout)
+		Expect(err).To(HaveOccurred(), "while waiting for pod to be running")
+
+		var podClient *framework.PodClient
+		podClient = f.PodClient()
+
+		By("updating the pod")
+		podClient.Update(podName, func(pod *v1.Pod) {
+			pod.ObjectMeta.Annotations = map[string]string{"mysubpath": "mypath"}
+		})
+
+		By("deleting the pod gracefully")
```

Comment:
this doesn't really check that the path got resolved correctly and the right file got surfaced in the pod, does it?

Reply:
Thinking about this some more, I think the subpath mount won't actually change if the env variable changes.  This is because subpath mount cleanup is asymmetrical with setup and they don't get cleaned up until pod termination.  So on a container restart, `makeMounts` will be called again, however, when we get to [prepareSubpathTarget](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/util/mount/nsenter_mount.go#L389:41), we find that the target (whose name doesn't depend on the subpath value), has already been mounted.

So the end behavior is that the subpath value cannot change on container restart.  Only new pods will get the new value.

Reply:
> So the end behavior is that the subpath value cannot change on container restart. Only new pods will get the new value.

once it succeeds, right? if an unresolvable value prevented successful setup, new values could be picked up on retry, correct?

Reply:
Correct. Once the subpath has been successfully setup, then it should not change for the lifetime of the pod.  Except... if the node reboots :-/